src/triage/component/catwalk/evaluation.py

```diff
@@ -101,23 +110,55 @@ def generate_binary_at_x(test_predictions, x_value, unit="top_n"):
     """Assign predicted classes based based on top% or absolute rank of score
 
     Args:
-        test_predictions (list) A list of predictions, sorted by risk desc
+        test_predictions (numpy.array) A predictions, sorted by risk score descending
         x_value (int) The percentile or absolute value desired
         unit (string, default 'top_n') The thresholding method desired,
             either percentile or top_n
 
-    Returns: (list) The predicted classes
+    Returns: (numpy.array) The predicted classes
     """
+    len_predictions = len(test_predictions)
+    if len_predictions == 0:
+        return numpy.array([])
     if unit == "percentile":
-        cutoff_index = int(len(test_predictions) * (x_value / 100.00))
+        cutoff_index = int(len_predictions * (x_value / 100.00))
     else:
-        cutoff_index = x_value
-    test_predictions_binary = [
-        1 if x < cutoff_index else 0 for x in range(len(test_predictions))
-    ]
+        cutoff_index = int(x_value)
+    num_ones = cutoff_index if cutoff_index <= len_predictions else len_predictions
+    num_zeroes = len_predictions - cutoff_index if cutoff_index <= len_predictions else 0
+    test_predictions_binary = numpy.concatenate(
+        (numpy.ones(num_ones, numpy.int8), numpy.zeros(num_zeroes, numpy.int8))
+    )
     return test_predictions_binary
 
 
+class MetricDefinition(
+    namedtuple('MetricDefinition', [
+        'metric',
+        'threshold_unit',
+        'threshold_value',
+        'parameter_combination',
+        'parameter_string'
+    ])
+):
+    """A single metric, bound to a particular threshold and parameter combination"""
+
+class MetricEvaluationResult(
+    namedtuple('MetricEvaluationResult', [
+        'metric',
+        'parameter',
+        'value',
+        'num_labeled_examples',
+        'num_labeled_above_threshold',
+        'num_positive_labels'
+    ])
+):
```

Comment:
I'm curious, how come subclass the `namedtuple`s? If it's just to add `__doc__`, I think you have better options in Python3, (and which don't require you to add `__slots__`, as you otherwise probably should here).

`typing.NamedTuple` provides a typed interface, (and requires annotations) [[docs]](https://docs.python.org/3/library/typing.html#typing.NamedTuple):

```
class MetricEvaluationResult(typing.NamedTuple):
    """I have a doc string."""
    metric: typing.Any
    parameter: str = 'boop'
    …
```

If that doesn't suit, in Python3, you can also set `__doc__` procedurally:

```
MetricEvaluationResult = namedtuple('MetricEvaluationResult', …)
MetricEvaluationResult.__doc__ = """…"""
```

And there are other niceties in Python3 now. I'm guessing [SimpleNamespace](https://docs.python.org/3/library/types.html#types.SimpleNamespace) doesn't fit the bill here; but, [dataclass](https://docs.python.org/3/library/dataclasses.html#module-dataclasses) very well might. `dataclass` is rather like `namedtuple`, but more `object`-y:

```
@dataclass
class MetricEvaluationResult:
    """my docs"""
    metric: typing.Any
    parameter: str = 'boop'
```

It works quite like `namedtuple`, but the underlying implementation is quite different, since it's dressing up your class definition, rather than generating a subclass of `tuple`: It generates the appropriate `__init__` for you, as well as all the other boring methods you probably want, like `__repr__` and `__eq__`. You can even pass the `frozen` flag to make a `dataclass` immutable, like a `namedtuple`.

Ultimately, a `namedtuple` might be lighter weight. But there are pros and cons :man_shrugging: 

Lots of stuff in Python3, man :boom:

Reply:
Thanks! Dataclass looks like the ticket

Reply:
(yes, it was just to add the docstring)

Reply:
:+1: